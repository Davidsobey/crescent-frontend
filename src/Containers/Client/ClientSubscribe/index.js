import React from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { Field, reduxForm } from 'redux-form';
import { MenuItem } from 'material-ui/Menu';

import Card from '../../../Components/Card';
import Select from '../../../Components/Select';
import Button from '../../../Components/Button';
import ClientActions from '../../../Actions/ClientActions';
import CourseActions from '../../../Actions/CourseActions';
import LinearProgress from '../../../Components/LinearProgress';

const validate = () => {
  const errors = {};

  return errors;
};
const required = value => value ? undefined : 'Required';
const number = value => value && isNaN(Number(value)) ? 'Must be a number' : undefined;

/* eslint-disable react/prefer-stateless-function */
class SubscriptionCreate extends React.Component {
  constructor(props) {
    super(props);
    this.props.dispatch(CourseActions.getAll());
    this.props.dispatch(ClientActions.getAll());
  }
  submit = (values) => {
    const subscription = Object.assign({}, values);
    this.props.dispatch(ClientActions.subscribe(subscription));
  };

  render() {
    const isValidTest = (test => Array.isArray(test.questionIds) ? test.questionIds.length : false);
    const isValidModule = (module => 
      (Array.isArray(module.moduleMaterialIds) ? module.moduleMaterialIds.length : false) && 
      (Array.isArray(module.testIds) ? module.tests.filter(isValidTest).length : false)
    );
    const isValidCourse = (course => Array.isArray(course.modules) ? course.modules.filter(isValidModule).length : false);

    return (
      <Card width="600px" title="Subscribe To Course">
        <form
          onSubmit={this.props.handleSubmit(this.submit)}
          noValidate
          autoComplete="off"
          className="centerForm"
        >
          <div>
            <div>
              {this.props.courses_loading ? (
                <div>
                  <LinearProgress color="secondary" />
                  Loading Courses
                </div>
              ) : (
                <Field name="courseID" label="Course Name" component={Select} validate={[ required ]}>
                  {this.props.courses
                  .filter(isValidCourse)
                  .map(course => (
                    <MenuItem value={course.id} key={course.id}>
                      {course.name}
                    </MenuItem>
                  ))}
                </Field>
              )}
            </div>
            <div>
              {this.props.clients_loading ? (
                <div>
                  <LinearProgress color="secondary" />
                  Loading Clients
                </div>
              ) : (
                <Field name="clientID" label="Client Name" component={Select} validate={[ required ]}>
                  {this.props.clients.map(client => (
                    <MenuItem value={client.id} key={client.id}>
                      {client.name}
                    </MenuItem>
                  ))}
                </Field>
              )}
            </div>
          </div>
          {this.props.client_subscribing ? (
            <div style={{width: '400px'}}>
              <LinearProgress color="secondary" />
              Subscribing Client To Course
            </div>
          ) : (
            <div className="formAlignRight">
              <Button
                className="buttonFormat"
                variant="raised"
                color="primary"
                type="submit"
              >
                Subscribe
              </Button>
            </div>
          )}
        </form>
      </Card>
    );
  }
}

SubscriptionCreate.propTypes = {
  dispatch: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  courses: PropTypes.array,
  courses_loading: PropTypes.bool,
  clients: PropTypes.array,
  clients_loading: PropTypes.bool,
  client_subscribing: PropTypes.bool,
};

const mapStateToProps = state => ({
  courses: state.CourseReducer.courses,
  courses_loading: state.CourseReducer.loading,
  clients: state.ClientReducer.clients,
  clients_loading: state.ClientReducer.loading,
  client_subscribing: state.ClientReducer.subscribing,
});

const withForm = reduxForm(
  {
    form: 'subscriptionCreate',
    validate,
  },
  SubscriptionCreate,
);

export default compose(
  connect(mapStateToProps),
  withForm,
)(SubscriptionCreate);
