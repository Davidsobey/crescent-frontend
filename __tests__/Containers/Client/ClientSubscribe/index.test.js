import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store'; // Smart components

// Component to be tested
import ClientSubscribe from '../../../../src/Containers/Client/ClientSubscribe/index.js';


const mockStore = configureStore();
const initialState = {
  ClientReducer: {
    clients: [],
    loading: false,
    subscribing: false,
  },
  CourseReducer: {
    courses: [],
    loading: false,
    unsubscribed_courses: [],
    unsubscribed_courses_loading: false,
  },
  LoginReducer: {
    user: {
      name: 'aaa',
    }
  },
};
const store = mockStore(initialState);

describe('<ClientSubscribe />', () => {
  describe('render()', () => {
    test('renders the component', () => {
      const wrapper = shallow(<ClientSubscribe store={store} />);
      const component = wrapper.dive();

      expect(toJson(component)).toMatchSnapshot();
    });
  });
});