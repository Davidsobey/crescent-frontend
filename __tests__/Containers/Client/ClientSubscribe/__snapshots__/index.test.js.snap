// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClientSubscribe /> render() renders the component 1`] = `
<Connect(Form(SubscriptionCreate))
  client_subscribing={false}
  clients={Array []}
  clients_loading={false}
  courses={Array []}
  courses_loading={false}
  destroyOnUnmount={true}
  dispatch={[Function]}
  enableReinitialize={false}
  forceUnregisterOnUnmount={false}
  form="subscriptionCreate"
  getFormState={[Function]}
  keepDirtyOnReinitialize={false}
  persistentSubmitErrors={false}
  pure={true}
  shouldAsyncValidate={[Function]}
  shouldError={[Function]}
  shouldValidate={[Function]}
  shouldWarn={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  touchOnBlur={true}
  touchOnChange={false}
  unsubscribed_courses={Array []}
  unsubscribed_courses_loading={false}
  updateUnregisteredFields={false}
  user={
    Object {
      "name": "aaa",
    }
  }
  validate={[Function]}
/>
`;
